pkg/operator/encryption/shared_encryption.go

```diff
@@ -0,0 +1,867 @@
+
+	// we have consistent and completely current read secrets, the next step is determining a consistent write key.
+	// To do this, we will choose the most current write key and use that to write the data
+	allWriteSecretsAsExpected := true
+	for gr, grState := range encryptionState {
+		if grState.writeSecret == nil || grState.writeSecret.Name != writeSecret.Name {
+			allWriteSecretsAsExpected = false
+			klog.V(4).Infof("%s does not have write secret %s", gr, writeSecret.Name)
+			break
+		}
+	}
+	// if our write secrets aren't all the same, update all the write secrets and wait for stability.
+	// We can move write keys before all the data has been migrated
+	if !allWriteSecretsAsExpected {
+		klog.V(4).Infof("not all write secrets in sync")
+		for gr := range encryptionState {
+			grState := encryptionState[gr]
+			grState.writeSecret = writeSecret
+			encryptionState[gr] = grState
+		}
+		return encryptionState
+	}
+
+	// at this point all of our read and write secrets are the same.  We need to inspect the write secret to
+	// ensure that it has all the expect resources listed as having been migrated.  If this is true, then we
+	// can prune the read secrets down to a single entry.
+
+	migratedResourceString := writeSecret.Annotations[encryptionSecretMigratedResources]
+	// if no migration has happened, we need to wait until it has
+	if len(migratedResourceString) == 0 {
+		klog.V(4).Infof("write secret %s has not been migrated", writeSecret.Name)
+		return encryptionState
+	}
+	migratedResources := &migratedGroupResources{}
+	// if we cannot read the data, wait until we can
+	if err := json.Unmarshal([]byte(migratedResourceString), migratedResources); err != nil {
+		klog.V(4).Infof("write secret %s has invalid migrated resource string: %v", writeSecret.Name, err)
+		return encryptionState
+	}
+
+	// get a list of all the resources we have, so that we can compare against the migrated keys annotation.
+	allResources := make([]schema.GroupResource, 0, len(encryptionState))
+	for gr := range encryptionState {
+		allResources = append(allResources, gr)
+	}
+
+	foundCount := 0
+	for _, expected := range allResources {
+		if migratedResources.hasResource(expected) {
+			foundCount++
+		}
+	}
+	// if we did not find migration indications for all resources, then just wait until we do
+	if foundCount != len(allResources) {
+		klog.V(4).Infof("write secret %s is missing migrated resources expected=%s, actual=%s", writeSecret.Name, allResources, migratedResources.Resources)
+		return encryptionState
+	}
+
+	// if we have migrated all of our resources, the next step is remove all unnecessary read keys.
+	// We only need the write key now
+	for gr := range encryptionState {
+		grState := encryptionState[gr]
+		grState.readSecrets = []*corev1.Secret{writeSecret}
+		encryptionState[gr] = grState
+	}
+	klog.V(4).Infof("write secret %s set as sole write key", writeSecret.Name)
+	return encryptionState
+}
+
+type migratedGroupResources struct {
```

Comment:
// migratedGroupResources is the data structured stored in the encryption.apiserver.operator.openshift.io/migrated-resources of a key secret.